components/global/core/io.cellery.observability.k8s.client/src/main/java/io/cellery/observability/k8s/client/ComponentsEventSource.java

```diff
@@ -0,0 +1,259 @@
+                                    cell.getMetadata().getCreationTimestamp()).getTime());
+                    attributes.put(Constants.ATTRIBUTE_ACTION, action.toString());
+                    attributes.put(LAST_KNOWN_ACTIVE_TIMESTAMP, 0L);
+
+                    // Checks and add HTTP as Ingress type
+                    addHTTPTypes(httpObjectsList, sourceEventListener, attributes, ingressList);
+                    // Checks and add TCP as Ingress type
+                    addTCPTypes(tcpObjectsList, sourceEventListener, attributes, ingressList);
+                    // Checks and add HTTP as Ingress type
+                    addWEBTypes(httpObjectsList, sourceEventListener, attributes, cell, ingressList);
+
+                } catch (ParseException e) {
+                    logger.error("Ignored cell change due to creation timestamp parse failure", e);
+                }
+            }
+
+            @Override
+            public void onClose(KubernetesClientException cause) {
+                if (cause == null) {
+                    if (logger.isDebugEnabled()) {
+                        logger.debug("Kubernetes cell watcher closed successfully");
+                    }
+                } else {
+                    logger.error("Kubernetes cell watcher closed with error", cause);
+                }
+            }
+        });
+        if (logger.isDebugEnabled()) {
+            logger.debug("Created cell watcher");
+        }
+    }
+
+    @Override
+    public void disconnect() {
+        if (k8sClient != null) {
+            k8sClient.close();
+            if (logger.isDebugEnabled()) {
+                logger.debug("Closed API server client");
+            }
+        }
+    }
+
+    @Override
+    public void destroy() {
+        // Do nothing
+    }
+
+    @Override
+    public void pause() {
+        // Do nothing
+    }
+
+    @Override
+    public void resume() {
+        // Do nothing
+    }
+
+    @Override
+    public Map<String, Object> currentState() {
+        return null;
+    }
+
+    @Override
+    public void restoreState(Map<String, Object> map) {
+        // Do nothing
+    }
+
+    /**
+     * This method checks if the ingress type is of HTTP and adds it to the relevant cell data
+     */
+    public void addHTTPTypes(List<HTTP> httpObjectsList, SourceEventListener sourceEventListener,
+                             Map<String, Object> attributes, IngressTypes ingressTypes) {
+        if (httpObjectsList != null) {
+            for (HTTP httpObj : httpObjectsList) {
+                if (httpObj.getBackend() != null && !httpObj.getBackend().isEmpty()) {
+                    ingressTypes.addType(HTTP);
+                    attributes.put(INGRESS_TYPES, StringUtils.join(ingressTypes.getTypes(), ','));
+                    attributes.put(Constants.ATTRIBUTE_COMPONENT, httpObj.getBackend());
+                    sourceEventListener.onEvent(attributes, new String[0]);
```

Comment:
Is this meant to be called within for loop? Ie, every http object you will be calling the source event listener, and you are not cleaning up the content in the attributes? In that case, the final object will have all elements of the attributes.

Reply:
Noted. I will update the logic.